library/bignum_core.h

```diff
@@ -250,15 +250,15 @@ mbedtls_mpi_uint mbedtls_mpi_core_mla( mbedtls_mpi_uint *X, size_t X_limbs,
 mbedtls_mpi_uint mbedtls_mpi_core_montmul_init( const mbedtls_mpi_uint *N );
 
 /**
- * \brief Montgomery multiplication: X = A * B * R^-1 mod N  (HAC 14.36)
+ * \brief Montgomery multiplication: X = A * B * R^-1 mod N (HAC 14.36)
+ *
+ * \p A and \p B must be in canonical form. That is, < \p N.
  *
  * \p X may be aliased to \p A or \p N, or even \p B (if \p AN_limbs ==
```

Comment:
Presumably
```suggestion
 * \p X may be aliased to \p A, or even \p B (if \p AN_limbs ==
```
The documentation of N states that it may not alias X. Our general rule is that outputs may not alias moduli.

Reply:
Well, the function does actually work if X aliases N - there are already test cases for that. So in this case the documentation for N is wrong.

Reply:
I think that the documentation should state what we are promising. The implementation might actually do more than that, but that shouldn't be a problem.

In this case, I think it makes sense not to promise that X can alias N as it doesn't make much sense. (eg. in the modulus structure, the value of the modulus is const. Also, the post condition on X is that it is less than N, which in the case of aliasing is impossible in a tight interpretation and still confusing when we try to interpret it well.)

Reply:
For an output parameter to alias to an input parameter means the storage of the input parameter can be overwritten to be used for the output - the input value doesn't matter (whereas for two input parameters to alias, their values must be identical).

As long as `core` is just used internally, I agree with you, that it's unlikely that we will want to alias `X` and `N`. However, there's nothing to stop a library user calling these functions, and they may have their own reasons to minimise memory use and overwrite `N` with the output value, so I see no harm in documenting that this is safe (and, of course, testing it!).

Given that a temporary working area is supplied to this function, this kind of "over promise" (beyond what we need and currently expect) is unlikely to constrain any future implementation.

Reply:
> there's nothing to stop a library user calling these functions

Yes, there is: it's an internal function. Only the legacy bignum API is public and must obey its promises until the next major version change.

But because it's an internal function, we can make high-reaching promises if we want: we can rescind them at any time, we just need to check that it doesn't break our own code.